data/src/main/java/com/microsoft/azure/kusto/data/ClientImpl.java

```diff
@@ -97,32 +81,69 @@ public ClientImpl(ConnectionStringBuilder csb, CloseableHttpClient httpClient, b
         clusterUrl = csb.getClusterUrl();
         aadAuthenticationHelper = clusterUrl.toLowerCase().startsWith(CloudInfo.LOCALHOST) ? null : TokenProviderFactory.createTokenProvider(csb, httpClient);
         clientDetails = new ClientDetails(csb.getApplicationNameForTracing(), csb.getUserNameForTracing(), csb.getClientVersionForTracing());
-        this.httpClient = httpClient;
-        this.leaveHttpClientOpen = leaveHttpClientOpen;
     }
 
     @Override
-    public KustoOperationResult execute(String command) throws DataServiceException, DataClientException {
-        return execute(DEFAULT_DATABASE_NAME, command);
+    public Mono<KustoOperationResult> executeQueryAsync(String database, String command, ClientRequestProperties properties) {
+        KustoRequest kr = new KustoRequest(command, database, properties);
+        return executeQueryAsync(kr);
     }
 
-    @Override
-    public KustoOperationResult execute(String database, String command) throws DataServiceException, DataClientException {
-        return execute(database, command, null);
+    Mono<KustoOperationResult> executeQueryAsync(@NotNull KustoRequest kr) {
+        if (kr.getCommandType() != CommandType.QUERY) {
+            kr.setCommandType(CommandType.QUERY);
```

Comment:
Maybe we should throw if someone wants a query but doesn't set the type to query?

Reply:
That would be a major behavior change compared with what you have currently. Command type has never been exposed to the user before. If we exposed the KustoRequest object they would have the capability easily, otherwise we would have to create an even more overloaded method with command type and the request is already very overloaded as it is. We have the execute method that takes commandtype but it is not public because it was never public before.

But at that point why do we have separate methods for executeQuery and executeMgmt? You had asked me specifically to deprecate the generic "execute" methods and separate methods into executeQuery and executeMgmt. If you call executeMgmt, does that not imply that you want to perform a management action? If you call executeQuery does that not imply you want to execute a query?

My recommendation would be either do the above or expose the request object so that we don't further inundate developers with even more overloaded methods, and make maintainance on our side even more difficult by forcing ourselves to method chain even more than we already are.